### PR TITLE
Domain classes for SDK 2.0

### PR DESCRIPTION
**Pull request by chazemar**

### CODE REVIEW

**Comment by chazemar** [comment]
update done

File: OCast/Core/Models/OCastDomainSettings.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+@objc
+public class SettingsKeyPressedCommand: OCastMessage {
+    public let key: String
+    public let code: String
+    public let ctrl: Bool
+    public let alt: Bool
+    public let shift: Bool
+    public let meta: Bool
+    public let location: Int
```
Comment (fsuc) [inline OCast/Core/Models/OCastDomainSettings.swift]: location is an enum (int): https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/location

File: OCast/Core/Models/OCastDomainWebApp.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//  Created by Christophe Azemar on 28/03/2019.
+//  Copyright © 2019 Orange. All rights reserved.
+//
+
+import Foundation
+
+// MARK: - WebApp Event
+@objc
+public class OCastWebAppConnectEvent: OCastMessage {
+    public let status: String
```
Comment (fsuc) [inline OCast/Core/Models/OCastDomainWebApp.swift]: enum : "connected" or "disconnected" and rename in OCastWebAppConnectedStatusEvent

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+//
+//  OCastDomainMedia.swift
+//  OCast
+//
+//  Created by Christophe Azemar on 28/03/2019.
+//  Copyright © 2019 Orange. All rights reserved.
+//
+
+import Foundation
+import DynamicCodable
```
Comment (fsuc) [inline OCast/Core/Models/OCastDomainMedia.swift]: to remove

File: OCast/Core/Models/OCastDomainGeneral.swift [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+//  Created by Christophe Azemar on 16/11/2018.
+//  Copyright © 2018 Orange. All rights reserved.
+//
+
+import Foundation
+import DynamicCodable
+
+public let OCastErrorDomain = "OCastError"
+
+public let kOCastPlaybackStatusEvent = "OCastPlaybackStatusEvent"
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainGeneral.swift]: use in future classes

File: OCast/Core/Models/OCastDomainWebApp.swift (hunk 2 above)
Comment (chazemar) [reply OCast/Core/Models/OCastDomainWebApp.swift]: done

File: OCast/Core/Models/OCastDomainGeneral.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+public typealias OCastMessage = NSObject & Codable
+
+public enum OCastDomainName: String {
+    case browser = "browser"
+    case settings = "settings"
+    case all = "*"
+}
+
+public class OCastGenericDeviceLayer<T: Codable>: OCastMessage {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainGeneral.swift]: updated

File: OCast/Core/Models/OCastDomainGeneral.swift [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    }
+    
+    public convenience init(source: String, destination: String, id: Int, status: String?, type: String, service: String, name: String, params: T, options: [String: Any]?) {
+        let reference = OCastGenericDataLayer(name: name, params: params, options: options)
+        let application = OCastGenericApplicationLayer(service: service, data: reference)
+        self.init(source: source, destination: destination, id: id, status: status, type: type, message: application)
+    }
+}
+
+public class OCastGenericApplicationLayer<T: Codable>: OCastMessage {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainGeneral.swift]: updated

File: OCast/Core/Models/OCastDomainGeneral.swift [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    public let service: String
+    public let data: OCastGenericDataLayer<T>
+    
+    public init(service: String, data: OCastGenericDataLayer<T>) {
+        self.service = service
+        self.data = data
+    }
+}
+
+public class OCastGenericDataLayer<T: Codable>: OCastMessage {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainGeneral.swift]: updated

File: OCast/Core/Models/OCastDomainGeneral.swift [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(name, forKey: .name)
+        try container.encode(params, forKey: .params)
+        try container.encodeIfPresent(options, forKey: .options)
+    }
+}
+
+@objc
+public class OCastDefaultDataParams: OCastMessage {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainGeneral.swift]: updated

File: OCast/Core/Models/OCastDomainMedia.swift (hunk 3 above)
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: removed

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+// MARK: - Media objects
+
+/**
+ Media type
+ - `.audio`: audio type
+ - `.image`: image type
+ - `.video`: viedo type
+ */
+@objc public enum OCastMediaType: Int {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: done

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+        switch (type) {
+        case "audio": self = .audio
+        case "video": self = .video
+        case "image": self = .image
+        default: self = .audio
+        }
+    }
+}
+
+@objc public enum OCastMediaTrackType: Int {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: done

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+public class MediaPlaybackStatus: OCastMessage {
+    public var volume: Float { return _volume ?? 0.0 }
+    public var mute: Bool { return _mute ?? false }
+    public var state: MediaPlaybackStatusState { return _state ?? .unknown }
+    public var position: Double { return _position ?? 0.0 }
+    public var duration: Double { return _duration ?? 0.0 }
+    
+    private let _volume: Float?
+    private let _mute: Bool?
+    private let _state: MediaPlaybackStatusState?
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: ok for this property. State is mandatory.

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+    public let videoTracks: [MediaTrack]
+    
+    enum CodingKeys : String, CodingKey {
+        case title = "title", subtitle = "subtitle", logo = "logo", _mediaType = "mediaType", textTracks = "textTracks", audioTracks = "audioTracks", videoTracks = "videoTracks"
+    }
+}
+
+@objc
+public class MediaTrack: OCastMessage {
+    public let language: String
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: added

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+    public let enable: Bool
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: updated

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+
+// MARK: - Media Commands
+@objc
+public class MediaPrepareCommand: OCastMessage {
+    public let url: String
+    public let frequency: UInt
+    public let title: String
+    public let subtitle: String
+    public let logo: String
+    public let mediaType: String
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: done

File: OCast/Core/Models/OCastDomainMedia.swift [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+public class MediaSeekCommand: OCastMessage {
+    public let position: Double
+    
+    public init(position: Double) {
+        self.position = position
+    }
+}
+
+@objc public class MediaGetPlaybackStatusCommand: OCastMessage {}
+@objc public class MediaGetDeviceIDCommand: OCastMessage {}
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainMedia.swift]: removed

File: OCast/Core/Models/OCastDomainSettings.swift [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+//
+//  Created by Christophe Azemar on 28/03/2019.
+//  Copyright © 2019 Orange. All rights reserved.
+//
+
+import Foundation
+
+// MARK: - Settings Objects
+@objc
+public class SettingsUpdateStatus: OCastMessage {
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainSettings.swift]: done

File: OCast/Core/Models/OCastDomainSettings.swift [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    public let state: String
+    public let version: String
+    public let progress: Int
```
[the remark this answers is not in the document]
Comment (chazemar) [reply OCast/Core/Models/OCastDomainSettings.swift]: legacy code : version optional, progress non optional

File: OCast/Core/Models/OCastDomainSettings.swift (hunk 1 above)
Comment (chazemar) [reply OCast/Core/Models/OCastDomainSettings.swift]: added

**Pull request merged**